crates/fluvio/src/producer/mod.rs

```diff
@@ -33,12 +45,224 @@ use crate::producer::partitioning::{Partitioner, SiphashRoundRobinPartitioner, P
+struct ProducerPool {
+    pool: Vec<Arc<PartitionProducer>>,
+}
+
+impl ProducerPool {
+    fn new(
+        topic: String,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>>,
+        linger: Duration,
+    ) -> Self {
+        let mut pool = vec![];
+        for (&partition_id, batch_list) in batches.iter() {
+            pool.push(PartitionProducer::shared(
+                topic.clone(),
+                partition_id,
+                spu_pool.clone(),
+                batch_list.clone(),
+                linger,
+            ));
+        }
+        Self { pool }
+    }
+
+    fn shared(
+        topic: String,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>>,
+        linger: Duration,
+    ) -> Arc<Self> {
+        Arc::new(ProducerPool::new(topic, spu_pool, batches, linger))
+    }
+}
+struct InnerTopicProducer {
+    topic: String,
+    spu_pool: Arc<SpuPool>,
+    partitioner: Box<dyn Partitioner + Send + Sync>,
+    record_accumulator: RecordAccumulator,
+    producer_pool: Arc<ProducerPool>,
+    linger: Duration,
+}
+
+impl InnerTopicProducer {
+    /// Flush all the PartitionProducers and wait for them.
+    async fn flush(&self, force: bool) -> Result<()> {
+        let mut futures = vec![];
+
+        for producer in self.producer_pool.pool.iter() {
+            let future = producer.flush(force);
+            futures.push(future);
+        }
+
+        join_all(futures)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(())
+    }
+
+    async fn push_record(self: Arc<Self>, record: Record) -> Result<PushRecord> {
+        let topics = self.spu_pool.metadata.topics();
+        let topic_spec = topics
+            .lookup_by_key(&self.topic)
+            .await?
+            .ok_or_else(|| FluvioError::TopicNotFound(self.topic.to_string()))?
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        let key = record.key.as_ref().map(|k| k.as_ref());
+        let value = record.value.as_ref();
+        let partition = self.partitioner.partition(&partition_config, key, value);
+
+        let push_record = self
+            .record_accumulator
+            .push_record(record, partition)
+            .await?;
+        let self_clone = self.clone();
+        if push_record.is_full {
+            fluvio_future::task::spawn(async move {
+                if let Err(e) = self_clone.flush(false).await {
+                    error!("Failed to flush producer: {:?}", e);
+                }
+            });
+        }
+
+        if push_record.new_batch {
+            let linger = self.linger;
+            fluvio_future::task::spawn(async move {
+                fluvio_future::timer::sleep(linger).await;
+                if let Err(e) = self.flush(false).await {
```

Comment:
flush doesn't need to be spawn

Reply:
Then how could I run the flush after the linger time is elapsed?

Reply:
there are 3 ways to flush:
* after duration
* when min size threshold has been reached
* manual flushing

so if the other 2 has been reached, lingering should be reset which implies use of `select`

Reply:
It is not clear to me how can I use select! here without creating a sender dispatcher